File: 3838-path-existence-queries-in-a-graph-i/path-existence-queries-in-a-graph-i.java
class Solution{
    int find(int node, int[] parent){
        if(node == parent[node])
            return node;

        return parent[node]=find(parent[node], parent);
    }

    void solve(int a, int b, int[] nums, int[] parent, int[] rank, int n){
        int x=find(a, parent);
        int y=find(b, parent);

        if(x == y)
            return;

        if(rank[x]<rank[y]){
            parent[x]=y;
            rank[y]++;
        }
        else{
            parent[y]=x;
            rank[x]++;
        }

    }

    public boolean[] pathExistenceQueries(int n, int[] nums, int mx, int[][] q){
        int[] parent=new int[n];
        int[] rank=new int[n];
        Arrays.fill(rank, 0);

        for(int i=0; i<n; i++){
            parent[i]=i;
        }
        
        for(int i=1; i<n; i++){
            if(nums[i]-nums[i-1]<=mx){
                solve(i-1, i, nums, parent, rank, n);
            }
        }

        boolean[] res=new boolean[q.length];
        Arrays.fill(res, false);
        for(int i=0; i<q.length; i++){
            int x=find(q[i][0], parent);
            int y=find(q[i][1], parent);

            if(x == y){
                res[i]=true;
            }
        }

        return res;

    }
}

File: Medium/3838-path-existence-queries-in-a-graph-i/path-existence-queries-in-a-graph-i.py
class Solution:
    def find(self, node, parent):
        if node == parent[node]:
            return node

        parent[node]=self.find(parent[node], parent)
        return parent[node]

    def solve(self, a, b, parent, rank):
        x=self.find(a, parent)
        y=self.find(b, parent)

        if x == y:
            return

        if rank[x]<rank[y]:
            parent[x]=y
            rank[y] += 1
        else:
            parent[y]=x
            rank[x] += 1
    
    def pathExistenceQueries(self, n: int, nums: List[int], mx: int, q: List[List[int]]) -> List[bool]:
        parent=list(range(n))
        rank=[0]*n
        
        for i in range(1, n):
            if nums[i]-nums[i-1]<=mx:
                self.solve(i-1, i, parent, rank)

        res=[False]*len(q)
        for i, (u, v) in enumerate(q):
            x=self.find(u, parent)
            y=self.find(v, parent)

            if x == y:
                res[i]=True

        return res